analyzers/src/SonarAnalyzer.Common/Rules/DateTimeFormatShouldNotBeHardcodedBase.cs

```diff
@@ -0,0 +1,67 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class DateTimeFormatShouldNotBeHardcodedBase<TSyntaxKind, TInvocation> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TInvocation : SyntaxNode
+{
+    private const string DiagnosticId = "S6585";
+    private const string ToStringLiteral = "ToString";
+
+    protected abstract Location InvalidArgumentLocation(TInvocation invocation);
+    protected abstract bool IsMultiCharStringLiteral(TInvocation invocation, SemanticModel semanticModel);
+
+    protected override string MessageFormat => "Do not hardcode the format specifier.";
+
+    protected IEnumerable<MemberDescriptor> CheckedMethods { get; } = new List<MemberDescriptor>
+        {
+            new(KnownType.System_DateTime, ToStringLiteral),
+            new(KnownType.System_DateTimeOffset, ToStringLiteral),
+            new(KnownType.System_DateOnly, ToStringLiteral),
+            new(KnownType.System_TimeOnly, ToStringLiteral),
+        };
+
+    protected DateTimeFormatShouldNotBeHardcodedBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, AnalyzeInvocation, Language.SyntaxKind.InvocationExpression);
+
+    private void AnalyzeInvocation(SonarSyntaxNodeReportingContext analysisContext)
+    {
+        if ((TInvocation)analysisContext.Node is var invocation
+            && Language.Syntax.InvocationIdentifier(invocation) is { } identifier
+            && CheckedMethods.Where(x => x.Name.Equals(identifier.ValueText, Language.NameComparison)) is var nameMatch
+            && nameMatch.Any()
+            && analysisContext.SemanticModel.GetSymbolInfo(identifier.Parent).Symbol is { } methodCallSymbol
+            && nameMatch.FirstOrDefault(x => methodCallSymbol.ContainingType.ConstructedFrom.Is(x.ContainingType)) is { }
+            && ShouldReportOnMethodCall(invocation, analysisContext.SemanticModel, methodCallSymbol))
+        {
+            analysisContext.ReportIssue(Diagnostic.Create(SupportedDiagnostics[0], InvalidArgumentLocation(invocation)));
+        }
+    }
+
+    private bool ShouldReportOnMethodCall(TInvocation invocation, SemanticModel semanticModel, ISymbol methodCallSymbol) =>
+        methodCallSymbol is IMethodSymbol methodSymbol
+        && methodSymbol.Parameters.Any()
+        && methodSymbol.Parameters[0].IsType(KnownType.System_String)
+        && IsMultiCharStringLiteral(invocation, semanticModel);
```

Comment:
It's worth adding a comment why it's important to only check for string literals longer than 1 character. It's not trivial, and one needs to check the MSDN documentation on formatting to know.